File: backend/routes/overview.py
from fastapi import APIRouter, HTTPException, Query
from datetime import date as date_cls, datetime, UTC
from typing import Optional
import calendar as calendar_module
from database import db
from models import CompletionToggle, Completion, compute_age_category
from services.season import is_in_winter_period, apply_winter_times

router = APIRouter(tags=["overview"])

DAY_NAMES = ["mandag", "tirsdag", "onsdag", "torsdag", "fredag", "lørdag", "søndag"]


@router.get("/daily-overview")
async def get_daily_overview(date: str = Query(...)):
    try:
        y, m, d = map(int, date.split("-"))
        weekday_idx = date_cls(y, m, d).weekday()
    except ValueError:
        raise HTTPException(status_code=400, detail="Ugyldigt datoformat, brug YYYY-MM-DD")
    day_of_week = DAY_NAMES[weekday_idx]

    dragons = await db.dragons.find().sort("created_at", 1).to_list(1000)
    times = await db.times.find().to_list(1000)
    times_map = {str(t["_id"]): t["time"] for t in times}
    settings_doc = await db.app_settings.find_one({"key": "app_settings_singleton"})
    language = (settings_doc or {}).get("language", "en")
    name_field = "name_da" if language == "da" else "name_en"
    task_items = await db.task_items.find().to_list(1000)
    items_map = {str(i["_id"]): (i.get(name_field) or i.get("name")) for i in task_items}
    completions = await db.completions.find({"date": date}).to_list(1000)
    completion_map = {(c["dragon_id"], c["schedule_slot_id"]): c["completed"] for c in completions}

    summer_start = (settings_doc or {}).get("light_summer_start", "03-01")
    winter_start = (settings_doc or {}).get("light_winter_start", "09-01")
    in_winter = is_in_winter_period(date_cls(y, m, d), summer_start, winter_start)
    times_winter_map = {str(t["_id"]): t.get("winter_time") for t in times}

    result = []
    for dragon in dragons:
        dragon_id = str(dragon["_id"])
        age_category = compute_age_category(dragon["birthday"])
        activity_state = dragon.get("activity_state", "active")
        slot_query = {
            "age_category": age_category,
            "day_of_week": day_of_week,
        }
        if activity_state == "brumation":
            # Bearded dragons in brumation stop eating - hide all feeding
            # tasks until switched back to "active". Care/Light&Heat tasks
            # (temperature checks, UVB, water, cleaning) still apply.
            slot_query["category"] = {"$ne": "fodring"}
        slots = await db.schedule_slots.find(slot_query).to_list(1000)
        tasks = []
        for slot in slots:
            slot_id = str(slot["_id"])
            time_str = times_map.get(slot["time_id"], "??:??")
            item_names = [items_map.get(iid) for iid in slot.get("item_ids", []) if items_map.get(iid)]
            tasks.append({
                "slot_id": slot_id,
                "time": time_str,
                "time_id": slot["time_id"],
                "category": slot["category"],
                "item_names": item_names,
                "is_automatic": slot.get("is_automatic", False),
                "completed": completion_map.get((dragon_id, slot_id), False),
            })
        tasks.sort(key=lambda t: t["time"])
        if in_winter:
            apply_winter_times(tasks, times_winter_map)
        for t in tasks:
            t.pop("time_id", None)
        result.append({
            "dragon_id": dragon_id,
            "name": dragon["name"],
            "photo_base64": dragon.get("photo_base64"),
            "age_category": age_category,
            "activity_state": activity_state,
            "tasks": tasks,
        })

    return {"date": date, "day_of_week": day_of_week, "is_winter_period": in_winter, "dragons": result}


@router.get("/completions/calendar-summary")
async def get_calendar_summary(
    year: int = Query(...),
    month: int = Query(..., ge=1, le=12),
    dragon_id: Optional[str] = Query(None),
):
    """Per-day completion status for a whole month, used by the Daily
    Overview's calendar date-picker. If dragon_id is given, status reflects
    only that dragon's tasks - otherwise it's aggregated across all dragons:
    - 'green'  = every manually-registrable task that day is done
    - 'yellow' = some but not all tasks are done
    - 'red'    = no tasks are done yet
    - 'none'   = no manually-registrable tasks were scheduled that day (edge case)
    Automatic (equipment) tasks like UVB/heat lamps are excluded from the
    count - they run on their own and aren't something the user 'registers'.
    """
    days_in_month = calendar_module.monthrange(year, month)[1]

    dragons = await db.dragons.find().to_list(1000)
    if dragon_id:
        dragons = [d for d in dragons if str(d["_id"]) == dragon_id]
    dragon_info = [
        {
            "id": str(d["_id"]),
            "age_category": compute_age_category(d["birthday"]),
            "activity_state": d.get("activity_state", "active"),
        }
        for d in dragons
    ]

    slots_cache: dict = {}

    async def get_slots(day_of_week: str, age_category: str, activity_state: str):
        key = (day_of_week, age_category, activity_state)
        if key not in slots_cache:
            query = {"age_category": age_category, "day_of_week": day_of_week}
            if activity_state == "brumation":
                query["category"] = {"$ne": "fodring"}
            slots_cache[key] = await db.schedule_slots.find(query).to_list(1000)
        return slots_cache[key]

    days_result = []
    for day_num in range(1, days_in_month + 1):
        d_obj = date_cls(year, month, day_num)
        date_str = d_obj.isoformat()
        day_of_week = DAY_NAMES[d_obj.weekday()]

        dragon_slots: dict = {}
        total = 0
        for info in dragon_info:
            slots = await get_slots(day_of_week, info["age_category"], info["activity_state"])
            # Only count manually-registrable tasks - automatic (equipment) tasks
            # like UVB/heat lamps run on their own and aren't something the user
            # 'registers', so they shouldn't affect the red/yellow/green status.
            registrable = [s for s in slots if not s.get("is_automatic")]
            dragon_slots[info["id"]] = registrable
            total += len(registrable)

        if total == 0:
            days_result.append({"date": date_str, "status": "none", "total": 0, "completed": 0})
            continue

        completions = await db.completions.find({"date": date_str}).to_list(1000)
        completion_map = {(c["dragon_id"], c["schedule_slot_id"]): c["completed"] for c in completions}

        completed = 0
        for info in dragon_info:
            for slot in dragon_slots[info["id"]]:
                slot_id = str(slot["_id"])
                if completion_map.get((info["id"], slot_id)):
                    completed += 1

        if completed == 0:
            status = "red"
        elif completed >= total:
            status = "green"
        else:
            status = "yellow"
        days_result.append({"date": date_str, "status": status, "total": total, "completed": completed})

    return {"year": year, "month": month, "days": days_result}


@router.post("/completions/toggle")
async def toggle_completion(payload: CompletionToggle):
    existing = await db.completions.find_one({
        "dragon_id": payload.dragon_id,
        "schedule_slot_id": payload.schedule_slot_id,
        "date": payload.date,
    })
    if existing:
        new_state = not existing["completed"]
        await db.completions.update_one(
            {"_id": existing["_id"]},
            {"$set": {"completed": new_state, "completed_at": datetime.now(UTC)}},
        )
        return {"completed": new_state}
    comp = Completion(
        dragon_id=payload.dragon_id,
        schedule_slot_id=payload.schedule_slot_id,
        date=payload.date,
        completed=True,
    )
    await db.completions.insert_one(comp.to_mongo())
    return {"completed": True}

File: frontend/localdb/overview.ts
/**
 * Daily overview aggregation + monthly calendar-summary - ported from the
 * previous backend's routes/overview.py, now running entirely on-device
 * against the local AsyncStorage-backed repositories.
 */
import { formatDateISO, getDanishDayOfWeek } from '@/constants/data';
import { getAll, KEYS } from './storage';
import { listDragons, listTimes, listTaskItems, getSettings, computeAgeCategory } from './repos';
import { isInWinterPeriod, applyWinterTimes, type WinterAdjustableTask } from './season';
import type { Completion, ScheduleSlot } from './types';

function completionKey(dragonId: string, slotId: string): string {
  return `${dragonId}\u0000${slotId}`;
}

export async function getDailyOverview(dateStr: string) {
  const [y, m, d] = dateStr.split('-').map((v) => parseInt(v, 10));
  const dateObj = new Date(y, m - 1, d);
  if (Number.isNaN(dateObj.getTime())) {
    throw new Error('Ugyldigt datoformat, brug YYYY-MM-DD');
  }
  const dayOfWeek = getDanishDayOfWeek(dateObj);

  const dragons = await listDragons();
  const times = await listTimes();
  const timesMap: Record<string, string> = {};
  const timesWinterMap: Record<string, string | null | undefined> = {};
  for (const t of times) {
    timesMap[t.id] = t.time;
    timesWinterMap[t.id] = t.winter_time;
  }

  const settings = await getSettings();
  const language = settings.language || 'en';
  const nameField = language === 'da' ? 'name_da' : 'name_en';

  const taskItems = await listTaskItems();
  const itemsMap: Record<string, string> = {};
  for (const item of taskItems) {
    const name = (item as any)[nameField] || item.name;
    if (name) itemsMap[item.id] = name;
  }

  const allSlots = await getAll<ScheduleSlot>(KEYS.SCHEDULE_SLOTS);
  const allCompletions = await getAll<Completion>(KEYS.COMPLETIONS);
  const dayCompletions = allCompletions.filter((c) => c.date === dateStr);
  const completionMap: Record<string, boolean> = {};
  for (const c of dayCompletions) {
    completionMap[completionKey(c.dragon_id, c.schedule_slot_id)] = c.completed;
  }

  const inWinter = isInWinterPeriod(dateObj, settings.light_summer_start, settings.light_winter_start);

  const result = [] as any[];
  for (const dragon of dragons) {
    const dragonId = dragon.id;
    const ageCategory = computeAgeCategory(dragon.birthday);
    const activityState = dragon.activity_state || 'active';

    let slots = allSlots.filter((s) => s.age_category === ageCategory && s.day_of_week === dayOfWeek);
    if (activityState === 'brumation') {
      // Bearded dragons in brumation stop eating - hide all feeding tasks
      // until switched back to "active". Care/Light&Heat tasks still apply.
      slots = slots.filter((s) => s.category !== 'fodring');
    }

    const tasks: (WinterAdjustableTask & { slot_id: string; item_names: string[]; completed: boolean })[] = slots.map((slot) => {
      const timeStr = timesMap[slot.time_id] || '??:??';
      const itemNames = slot.item_ids.map((iid) => itemsMap[iid]).filter(Boolean);
      return {
        slot_id: slot.id,
        time: timeStr,
        time_id: slot.time_id,
        category: slot.category,
        item_names: itemNames,
        is_automatic: slot.is_automatic,
        completed: completionMap[completionKey(dragonId, slot.id)] || false,
      };
    });
    tasks.sort((a, b) => (a.time < b.time ? -1 : a.time > b.time ? 1 : 0));

    if (inWinter) {
      applyWinterTimes(tasks, timesWinterMap);
    }
    const cleanTasks = tasks.map(({ time_id, ...rest }) => rest);

    result.push({
      dragon_id: dragonId,
      name: dragon.name,
      photo_base64: dragon.photo_base64,
      age_category: ageCategory,
      activity_state: activityState,
      tasks: cleanTasks,
    });
  }

  return { date: dateStr, day_of_week: dayOfWeek, is_winter_period: inWinter, dragons: result };
}

export async function getCalendarSummary(year: number, month: number, dragonId?: string | null) {
  const daysInMonth = new Date(year, month, 0).getDate();

  let dragons = await listDragons();
  if (dragonId) dragons = dragons.filter((d) => d.id === dragonId);
  const dragonInfo = dragons.map((d) => ({
    id: d.id,
    age_category: computeAgeCategory(d.birthday),
    activity_state: d.activity_state || 'active',
  }));

  const allSlots = await getAll<ScheduleSlot>(KEYS.SCHEDULE_SLOTS);
  const allCompletions = await getAll<Completion>(KEYS.COMPLETIONS);

  const slotsCache: Record<string, ScheduleSlot[]> = {};
  function getSlots(dayOfWeek: string, ageCategory: string, activityState: string): ScheduleSlot[] {
    const key = `${dayOfWeek}\u0000${ageCategory}\u0000${activityState}`;
    if (!slotsCache[key]) {
      let slots = allSlots.filter((s) => s.age_category === ageCategory && s.day_of_week === dayOfWeek);
      if (activityState === 'brumation') slots = slots.filter((s) => s.category !== 'fodring');
      slotsCache[key] = slots;
    }
    return slotsCache[key];
  }

  const daysResult: any[] = [];
  for (let dayNum = 1; dayNum <= daysInMonth; dayNum++) {
    const dObj = new Date(year, month - 1, dayNum);
    const dateStr = formatDateISO(dObj);
    const dayOfWeek = getDanishDayOfWeek(dObj);

    const dragonSlots: Record<string, ScheduleSlot[]> = {};
    let total = 0;
    for (const info of dragonInfo) {
      const slots = getSlots(dayOfWeek, info.age_category, info.activity_state);
      // Only count manually-registrable tasks - automatic (equipment) tasks
      // like UVB/heat lamps run on their own and shouldn't affect status.
      const registrable = slots.filter((s) => !s.is_automatic);
      dragonSlots[info.id] = registrable;
      total += registrable.length;
    }

    if (total === 0) {
      daysResult.push({ date: dateStr, status: 'none', total: 0, completed: 0 });
      continue;
    }

    const dayCompletions = allCompletions.filter((c) => c.date === dateStr);
    const completionMap: Record<string, boolean> = {};
    for (const c of dayCompletions) completionMap[completionKey(c.dragon_id, c.schedule_slot_id)] = c.completed;

    let completed = 0;
    for (const info of dragonInfo) {
      for (const slot of dragonSlots[info.id]) {
        if (completionMap[completionKey(info.id, slot.id)]) completed += 1;
      }
    }

    let status: 'green' | 'yellow' | 'red';
    if (completed === 0) status = 'red';
    else if (completed >= total) status = 'green';
    else status = 'yellow';

    daysResult.push({ date: dateStr, status, total, completed });
  }

  return { year, month, days: daysResult };
}
